### PR TITLE
Hue motion sensor SML001 improved

### PR DESCRIPTION
**Pull request by allofmex**
This is a fix for Hue motion sensor SML001.
The previous implementation has a serious problem: It fails if there is continuous motion.

Since there was a commit addressing an issue before already, here a summery (solution below):

The original implementation before  [this commit](https://github.com/Koenkk/zigbee2mqtt/issues/467) ignored the reported value and published "motion true" on any event, even if the sensor report said "motion false". 

Resulting in:
- Time X: Single Motion, Sensor sends "occupancy: 1" -> publish "Motion true", occupancy_timeout to 90 seconds	(CORRECT)
- X+10sec: Sensor sends "occupancy: 0" -> converter IGNORED value "0", published "Motion true" (WRONG), occupancy_timeout got reset to 90 seconds	
- X+10+90: occupancy_timeout occures, publish "Motion false" by converter	(more or less CORRECT)
- every 600sec: Sensors reports current status, "occupancy 0", value "0" IGNORED again, converter published "Motion true" 	(WRONG)
(Thats the 600 in devices.js "(cb) => device.report('msOccupancySensing', 'occupancy', 0, 600, null, cb),")

After the mentioned commit:
-  ...
- X+10sec: Sensor sends "occupancy 0" -> converter ignores all except value "1" -> nothing published or reset (WRONG)
- X+90sec: occupancy_timeout occures, publish "Motion false" by converter
- every 600sec: Sensor reports "occupancy 0", ignored because not "1", nothing published, issue seems solved, because no wrong events even if sensor stored in box...

Now the current issue:
- ...
- X+x+xsec: continous motion, sensor resets every time its internal timer. No attrReports are send because state does not change (still occupancy 1/true)
- X+90sec: occupancy_timeout in converter occures -> publish "Motion false" 	(WRONG, sensor is still on occupancy true)
- The timeout gets only reset if there is no motion for > 10 sec, because Sensor changes to "0" (report ignored by converter) and at next motion to "1" (report accepted, timeout reset)

**Solution:**
No occupancy_timeout timer in converter! We just handle the reported value.
The hue sensor sends both messages, occupancy true (motion detected) and false after no motion was detected for 10 seconds (default setting). So we don't need the occupancy_timeout timer.

I made lots of tests, the motion start/stop messages are correctly reported by the sensor. And if the stop message may be missed once (lost packet), the next periodic report (the 600 sec) will publish correct value. (I changed this value to 90% of occupancy_timeout, just in case...)

Here a log of the occupancy related events the sensor is generating (pirOToUDelay: 0, reporting max intervall: 600)

> 2018-12-31T07:11:58.423Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 1 } ]
> 2018-12-31T07:11:58.424Z   ind:changed                                  { cid: 'msOccupancySensing', data: { occupancy: 1 } }
> 2018-12-31T07:12:00.043Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 1 } ]
> 2018-12-31T07:12:08.424Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:12:08.424Z   ind:changed                                  { cid: 'msOccupancySensing', data: { occupancy: 0 } }
> 2018-12-31T07:12:10.019Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:22:08.174Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:22:09.771Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:32:08.086Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:32:09.689Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:42:07.967Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:42:09.565Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:52:07.877Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T07:52:09.477Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:02:07.695Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:02:09.297Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:12:07.388Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:12:08.985Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> ...
> 2018-12-31T08:31:23.790Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:32:46.571Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 1 } ]
> 2018-12-31T08:32:46.572Z   ind:changed                                  { cid: 'msOccupancySensing', data: { occupancy: 1 } }
> 2018-12-31T08:32:48.184Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 1 } ]
> 2018-12-31T08:36:06.075Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]
> 2018-12-31T08:36:06.077Z   ind:changed                                  { cid: 'msOccupancySensing', data: { occupancy: 0 } }
> 2018-12-31T08:36:07.701Z   ind:reported   msOccupancySensing            [ { attrId: 0, dataType: 24, attrData: 0 } ]

- motion start is reported
- after 10 seconds motion stop is reported
- every 600 seconds a current status report (twice)
- the last event was 3 minutes permanent motion in front of sensor

(only ind:reported is used by converter)

Note: An "occupancy timeout" can be configured by setting sensors pirOToUDelay value. I worked on this already, will be submited in second pull request.

I moved the original converter implementation to generic_occupancy_no_off_msg, so behavior for other sensors will not change.

### CODE REVIEW

**Comment by simonvanderveldt** [comment]
:+1: 

I just wanted to create an issue for this since I'm also observing this behavior and it seems incorrect to me.
Just did a bunch of tests and my observations match yours:

Trigger motion once:
```
t=0 attReport + devChange occupancy 1
t+2 another attReport occupancy 1 (not sure why this one happens)
t+10 attReport + devChange occupancy 0
t+12 another attReport occupancy 0 (again not sure why this one happens)
```
Trigger motion continuously for 60s:
```
t=0 attReport + devChange occupancy 1
t+2 another attReport occupancy 1 (not sure why this one happens)
t+70 (i.e. 60s motion + 10s delay from the sensor itself) attReport + devChange occupancy 0
t+72 another attReport occupancy 0 (again not sure why this one happens)
````

So what the sensor does is the following:
```
- Detect motion
- Send occupancy 1
- (If there's continuous motion) wait until there's no motion detected
- Wait 10s
- Send occupancy 0
```

This behavior from the device combined with the behavior from zigbee2mqtt means that it's currently impossible to handle continuous motion (which means the lights turn off even when I'm still in the room).

I like the proposed solution, no abstractions, just forward the value as reported by the sensor. Very much prefer this over having logic around this in zigbee2mqtt which I see as a simple conversion that shouldn't contain any logic. This way everyone can add their desired logic around this themselves in their chosen solution.

I do believe @Koenkk mentioned the `occupancy_false` reporting always happening was by design, so maybe he has a different opinion on it.

File: devices.js [hunk 1]
```diff
@@ -744,7 +744,13 @@ const devices = [
...
                 (cb) => device.bind('msOccupancySensing', coordinator, cb),
                 (cb) => device.report('genPowerCfg', 'batteryPercentageRemaining', 0, 1000, 0, cb),
-                (cb) => device.report('msOccupancySensing', 'occupancy', 0, 600, null, cb),
+                // Device should report at least short before occupancyTimeout.
+                // Warning: in case options occupancy_timeout is used instead of default
+                // occupancyTimeout, max reporting intervall needs to be set to an interval
+                // shorter than occupancy_timeout. Else our timer may timeout before
+                // device reports its status. (important in case of continous motion)
+                (cb) => device.report('msOccupancySensing', 'occupancy',
+                    0, Math.round(fz.occupancyTimeout * 0.9), null, cb),
```
Comment (Koenkk) [inline devices.js]: AFAIK the occupancyTimeout is not exported, how can this work?

**Comment by Koenkk** [comment]
`occupancy` `false` should only be generated for motion sensors that do not send it, meaning this implementation is good. 

I've the comment above is fixed this can be merged.

File: devices.js (hunk 1 above)
Comment (allofmex) [reply devices.js]: done

**Comment by sti0** [comment]
@Koenkk sorry for my question, but how can I use this fix? I'm using the HASSIO addon. Have I to wait for an update?

**Comment by Koenkk** [comment]
@sti0 you can use the hassio `-edge` version.

**Pull request merged**